pkg/sync/task/future/future.go

```diff
@@ -0,0 +1,67 @@
+package future
+
+import "context"
+
+type Value interface{}
+
+type SetFunc func(Value)
+type SetErrorFunc func(error)
+
+// Thread-safe facility to obtain/wait for a value from a single producer in
+// multiple goroutines
+type Future struct {
+	value Value
+	err   error
+
+	ready chan struct{}
+}
+
+// New constructs new Future.
+func New() (*Future, SetFunc, SetErrorFunc) {
+	f := &Future{ready: make(chan struct{})}
+	return f, f.set, f.setError
+}
+
+// Get returns value when it is set.
+func (f *Future) Get(ctx context.Context) (Value, error) {
+	if f.Ready() {
+		return f.value, f.err
+	}
+
+	select {
+	case <-f.ready:
+		return f.value, f.err
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+// Set value of the future and change state to ready.
+func (f *Future) set(value Value) {
+	select {
+	case <-f.ready:
+	default:
+		f.value = value
+		close(f.ready)
+	}
```

Comment:
this is not threadsafe -> if 2 goroutines are called at the same time both of them can be at line 44

Reply:
set is returned from `New` only, so it's accessible in the goroutine it is constructed in

Reply:
Hmm, `oauth2.Token` is a value type, so I don't actually need this type, I can just use channels to wait for data and channels are always thread-safe, so that would be better.